src/producer/TransactionMQProducer.cpp

```diff
@@ -0,0 +1,218 @@
+  } catch (MQException& e) {
+    THROW_MQEXCEPTION(MQClientException, e.what(), -1);
+  }
+
+  LocalTransactionState localTransactionState = LocalTransactionState::UNKNOW;
+  switch (sendResult.getSendStatus()) {
+    case SendStatus::SEND_OK:
+      try {
+        if (sendResult.getTransactionId() != "") {
+          msg.setProperty("__transactionId__", sendResult.getTransactionId());
+        }
+        string transactionId = msg.getProperty(MQMessage::PROPERTY_UNIQ_CLIENT_MESSAGE_ID_KEYIDX);
+        if (transactionId != "") {
+          msg.setTransactionId(transactionId);
+        }
+        LOG_DEBUG("sendMessageInTransaction, msgId:%s, transactionId:%s", sendResult.getMsgId().data(),
+                  transactionId.data());
+        localTransactionState = m_transactionListener->executeLocalTransaction(msg, arg);
+        if (localTransactionState != LocalTransactionState::COMMIT_MESSAGE) {
+          LOG_WARN("executeLocalTransaction ret not LocalTransactionState::commit, msg:%s", msg.toString().data());
+        }
+      } catch (MQException& e) {
+        THROW_MQEXCEPTION(MQClientException, e.what(), -1);
+      }
+      break;
+    case SendStatus::SEND_FLUSH_DISK_TIMEOUT:
+    case SendStatus::SEND_FLUSH_SLAVE_TIMEOUT:
+    case SendStatus::SEND_SLAVE_NOT_AVAILABLE:
+      localTransactionState = LocalTransactionState::ROLLBACK_MESSAGE;
+      LOG_WARN("sendMessageInTransaction, send not ok, rollback, result:%s", sendResult.toString().data());
+      break;
+    default:
+      break;
+  }
+
+  try {
+    endTransaction(sendResult, localTransactionState);
+  } catch (MQException& e) {
+    LOG_WARN("endTransaction exception:%s", e.what());
+  }
+
+  TransactionSendResult transactionSendResult(sendResult.getSendStatus(), sendResult.getMsgId(),
+                                              sendResult.getOffsetMsgId(), sendResult.getMessageQueue(),
+                                              sendResult.getQueueOffset());
+  transactionSendResult.setTransactionId(msg.getTransactionId());
+  transactionSendResult.setLocalTransactionState(localTransactionState);
+  return transactionSendResult;
+}
+
+void TransactionMQProducer::endTransaction(SendResult& sendResult, LocalTransactionState& localTransactionState) {
+  MQMessageId id;
+  if (sendResult.getOffsetMsgId() != "") {
+    id = MQDecoder::decodeMessageId(sendResult.getOffsetMsgId());
+  } else {
+    id = MQDecoder::decodeMessageId(sendResult.getMsgId());
+  }
+  string transId = sendResult.getTransactionId();
+
+  int commitOrRollback = 0;
+  switch (localTransactionState) {
+    case COMMIT_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionCommitType;
+      break;
+    case ROLLBACK_MESSAGE:
+      commitOrRollback = MessageSysFlag::TransactionRollbackType;
+      break;
+    case UNKNOW:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
+      break;
+    default:
+      commitOrRollback = MessageSysFlag::TransactionNotType;
```

Comment:
default case don't need do anything, because enum LocalTransactionState only have three. but this will defend hack possible.

Reply:
changed